pkg/reconciler/providerconfig/reconciler.go

```diff
@@ -0,0 +1,198 @@
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(l logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = l
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// NewReconciler returns a Reconciler of ProviderConfigs.
+func NewReconciler(m manager.Manager, of resource.ProviderConfigKinds, o ...ReconcilerOption) *Reconciler {
+	nc := func() resource.ProviderConfig {
+		return resource.MustCreateObject(of.Config, m.GetScheme()).(resource.ProviderConfig)
+	}
+	nul := func() resource.ProviderConfigUsageList {
+		return resource.MustCreateObject(of.UsageList, m.GetScheme()).(resource.ProviderConfigUsageList)
+	}
+
+	// Panic early if we've been asked to reconcile a resource kind that has not
+	// been registered with our controller manager's scheme.
+	_, _ = nc(), nul()
+
+	r := &Reconciler{
+		client: m.GetClient(),
+
+		newConfig:    nc,
+		newUsageList: nul,
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// Reconcile a ProviderConfig by accounting for the managed resources that are
+// using it, and ensuring it cannot be deleted until it is no longer in use.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	pc := r.newConfig()
+	if err := r.client.Get(ctx, req.NamespacedName, pc); err != nil {
+		// In case object is not found, most likely the object was deleted and
+		// then disappeared while the event was in the processing queue. We
+		// don't need to take any action in that case.
+		log.Debug(errGetPC, "error", err)
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetPC)
+	}
+
+	log = log.WithValues(
+		"uid", pc.GetUID(),
+		"version", pc.GetResourceVersion(),
+		"name", pc.GetName(),
+	)
+
+	l := r.newUsageList()
+	if err := r.client.List(ctx, l, client.MatchingLabels{v1alpha1.LabelKeyProviderName: pc.GetName()}); err != nil {
+		log.Debug(errListPCUs, "error", err)
+		r.record.Event(pc, event.Warning(reasonAccount, errors.Wrap(err, errListPCUs)))
+		return reconcile.Result{RequeueAfter: shortWait}, nil
+	}
+
+	users := int64(len(l.GetItems()))
+	for _, pcu := range l.GetItems() {
+		if metav1.GetControllerOf(pcu) == nil {
+			// Usages should always have a controller reference. If this one has
+			// none it's probably been stripped off (e.g. by a Velero restore).
+			// We can safely delete it - it's either stale, or will be recreated
+			// next time the relevant managed resource connects.
+			if err := r.client.Delete(ctx, pcu); resource.IgnoreNotFound(err) != nil {
```

Comment:
What happens if `spec.providerConfigRef` is changed to another value? Checking with https://github.com/crossplane/provider-gcp/pull/262 , it seems this would block deletion of the old `ProviderConfig` without any reason. Should we make sure the object pointed with resource reference in PCU is still using that PC? If not, delete the PCU?

Reply:
The PCU is updated on every managed resource reconcile loop if it has changed from what is in cache. So if the managed resource's `spec.providerConfigRef` changes:

* The PCU's `providerConfigRef` also changes to reflect the new name.
* The PCU's `crossplane.io/provider-config` label changes to reflect the new name.
* The PCU would no longer be included in the `List` call above for the old PC.

I think the only risk of a PCU getting stuck pointing to an old PC is if:

* The managed resource's `spec.providerConfigRef` is updated.
* The managed resource fails to get the new PC, or its secret, and thus never writes its new PCU.

I think a reasonable way to avoid this that would not involve having to get and validate the managed resource would be to simply write the PCU _before_ attempting to get and use the PC. i.e. The PCU declares the intent to use a PC, rather than a successful use of the PC.

That all said, I'm skeptical that changing the `spec.providerConfigRef` of a managed resource is something we should allow. It seems like it could easily result in a managed resource becoming orphaned or broken. I'm not immediately sure what reason someone would have to change what PC a managed resource was using.

Reply:
> I'm skeptical that changing the spec.providerConfigRef of a managed resource is something we should allow. It seems like it could easily result in a managed resource becoming orphaned or broken. I'm not immediately sure what reason someone would have to change what PC a managed resource was using.

I can't see big use cases, too. But I wouldn't be able explain why a user wouldn't be able to change if they wanted to, either. Normally, the resource is there and cloud providers allow you to manipulate it with any creds you want as long as they got the necessary permissions. Diverging from that doesn't really feel right.

> to simply write the PCU before attempting to get and use the PC. i.e. The PCU declares the intent to use a PC, rather than a successful use of the PC.

This sounds good. Though I wouldn't frame PCU as the intent to use. In the end, we don't reject the change on `spec.providerConfigRef` and that's the ultimate source of truth, i.e. we don't use PCU to decide what `ProviderConfig` to use during `Connect`. It's mostly machinery and not being able to reconcile the managed resource if PCU couldn't be updated looks like a negative side effect of that machinery, though I don't expect that to happen frequently anyway.

Reply:
> Though I wouldn't frame PCU as the intent to use.

I didn't mean 'declares' in the sense that it's the declarative source of truth. 'Indicates' might have been clearer - i.e. a PCU indicates that a resource is trying or intending to use a PC, not that it has successfully done so.